File: content/snippets/python/language_basics_trees_and_graphs.py
#!/usr/bin/env python3
from __future__ import annotations
from collections import deque
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TreeNode:
    value: int
    left: Optional[TreeNode] = field(default=None, repr=False)
    right: Optional[TreeNode] = field(default=None, repr=False)


def insert(root: Optional[TreeNode], value: int) -> TreeNode:
    if root is None:
        return TreeNode(value)
    if value < root.value:
        root.left = insert(root.left, value)
    else:
        root.right = insert(root.right, value)
    return root


def in_order(root: Optional[TreeNode]) -> list[int]:
    if root is None:
        return []
    return in_order(root.left) + [root.value] + in_order(root.right)


def bfs(root: Optional[TreeNode]) -> list[int]:
    if root is None:
        return []
    result: list[int] = []
    queue: deque[TreeNode] = deque([root])
    while queue:
        node = queue.popleft()
        result.append(node.value)
        if node.left:  queue.append(node.left)
        if node.right: queue.append(node.right)
    return result


def main() -> None:
    root: Optional[TreeNode] = None
    for v in [5, 3, 7, 1, 4]:
        root = insert(root, v)
    print("in-order:", in_order(root))  # [1, 3, 4, 5, 7]
    print("bfs:     ", bfs(root))       # [5, 3, 7, 1, 4]


if __name__ == "__main__":
    main()

File: content/snippets/typescript/language_basics_trees_and_graphs.ts
#!/usr/bin/env ts-node

interface TreeNode {
  value: number;
  left: TreeNode | null;
  right: TreeNode | null;
}

function makeNode(value: number): TreeNode {
  return { value, left: null, right: null };
}

function insert(root: TreeNode | null, value: number): TreeNode {
  if (root === null) return makeNode(value);
  if (value < root.value) root.left  = insert(root.left,  value);
  else                    root.right = insert(root.right, value);
  return root;
}

function inOrder(root: TreeNode | null): number[] {
  if (root === null) return [];
  return [...inOrder(root.left), root.value, ...inOrder(root.right)];
}

function bfs(root: TreeNode | null): number[] {
  if (root === null) return [];
  const result: number[] = [];
  const queue: TreeNode[] = [root];
  while (queue.length > 0) {
    const node = queue.shift()!;
    result.push(node.value);
    if (node.left)  queue.push(node.left);
    if (node.right) queue.push(node.right);
  }
  return result;
}

function main(): void {
  let root: TreeNode | null = null;
  for (const v of [5, 3, 7, 1, 4]) root = insert(root, v);
  console.log("in-order:", inOrder(root));  // [1, 3, 4, 5, 7]
  console.log("bfs:     ", bfs(root));      // [5, 3, 7, 1, 4]
}

main();
